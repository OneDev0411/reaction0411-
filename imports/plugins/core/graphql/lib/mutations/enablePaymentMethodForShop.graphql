enablePaymentMethodForShop(input: EnablePaymentMethodForShopInput!) {
  enablePaymentMethodForShop(input: $input) {
    method {
      name
      displayName
      isEnabled
    }
  }
}
