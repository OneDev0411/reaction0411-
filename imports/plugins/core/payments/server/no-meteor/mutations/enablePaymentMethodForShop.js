import ReactionError from "@reactioncommerce/reaction-error";
import SimpleSchema from "simpl-schema";
import { Shops } from "/lib/collections";
import { paymentMethods as allPaymentMethods } from "/imports/plugins/core/core/server/no-meteor/pluginRegistration";

const paramsSchema = new SimpleSchema({
  isEnabled: Boolean,
  paymentMethodName: String,
  shopId: String
});

export default async function enablePaymentMethodForShop(context, input = {}) {
  paramsSchema.validate(input);
  const { isEnabled, paymentMethodName, shopId } = input;

  if (!allPaymentMethods[paymentMethodName]) {
    throw new ReactionError("not-found", "Requested payment method is invalid");
  }

  const shop = await context.queries.shopById(context, shopId);
  const methods = new Set(shop.availablePaymentMethods);

  if (isEnabled) {
    methods.add(paymentMethodName);
  } else {
    methods.delete(paymentMethodName);
  }

  await Shops.update(
    { _id: shop._id },
    { $set: { availablePaymentMethods: Array.from(methods) } },
    { bypassCollection2: true }
  );

  return context.queries.paymentMethods(context, shopId);
}
