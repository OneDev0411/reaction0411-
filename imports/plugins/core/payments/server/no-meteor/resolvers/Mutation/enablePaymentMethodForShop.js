import { decodeShopOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/shop";

export default async function enablePaymentMethodForShop(parentResult, { input }, context) {
  input.shopId = decodeShopOpaqueId(input.shopId);
  return context.mutations.enablePaymentMethodForShop(context, input);
}
