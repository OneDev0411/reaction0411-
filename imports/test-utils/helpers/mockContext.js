const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getAbsoluteUrl: jest.fn().mockName("getAbsoluteUrl").mockImplementation((path) => {
    const adjustedPath = path[0] === "/" ? path : `/${path}`;
    return `https://app.mock${adjustedPath}`;
  }),
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  mutations: {},
  queries: {},
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

/**
 * @summary Returns a mock collection instance with the given name
 * @param {String} collectionName The collection name
 * @returns {Object} Mock collection instance
 */
export function mockCollection(collectionName) {
  return {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    bulkWrite: jest.fn().mockName(`${collectionName}.bulkWrite`).mockReturnValue(Promise.resolve({
      nMatched: 2,
      nModified: 2,
      result: {
        writeErrors: []
      }
    })),
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    findOneAndDelete: jest.fn().mockName(`${collectionName}.findOneAndDelete`),
    findOneAndUpdate: jest.fn().mockName(`${collectionName}.findOneAndUpdate`),
    fetch: jest.fn().mockName(`${collectionName}.fetch`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
}

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "ExampleIOUPaymentRefunds",
  "Groups",
  "MediaRecords",
  "NavigationItems",
  "NavigationTrees",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "SimpleInventory",
  "Tags",
  "Templates",
  "Themes",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = mockCollection(collectionName);
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
