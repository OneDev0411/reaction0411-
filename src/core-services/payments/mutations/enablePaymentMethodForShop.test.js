import mockContext from "@reactioncommerce/api-utils/tests/mockContext.js";
import Factory from "/tests/util/factory.js"; // TODO: research how to add `factory.js` to `api-utils` (https://github.com/reactioncommerce/reaction/issues/5646)
import enablePaymentMethodForShop from "./enablePaymentMethodForShop.js";

jest.mock("../registration", () => ({
  paymentMethods: {
    mockPaymentMethod: {
      name: "mockPaymentMethod",
      displayName: "Mock!",
      pluginName: "mock-plugin"
    }
  }
}));

const fakeShop = Factory.Shop.makeOne();
const mockEnablePaymentMethod = jest.fn().mockName("enablePaymentMethodForShop");
const mockPaymentMethods = jest.fn().mockName("paymentMethods");
const mockShopById = jest.fn().mockName("shopById");

beforeAll(() => {
  mockContext.queries = {
    paymentMethods: mockPaymentMethods,
    shopById: mockShopById
  };
  mockContext.mutations = { enablePaymentMethodForShop: mockEnablePaymentMethod };
});

beforeEach(() => {
  jest.resetAllMocks();
  mockShopById.mockClear();
  mockEnablePaymentMethod.mockClear();
  fakeShop.availablePaymentMethods = [];
});

test("throws if userHasPermissionLegacy returns false", async () => {
  mockContext.userHasPermissionLegacy.mockReturnValue(false);
  mockShopById.mockReturnValue(fakeShop);

  await expect(enablePaymentMethodForShop(mockContext, {
    shopId: fakeShop._id,
    paymentMethodName: "does not exist",
    isEnabled: true
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("errors on missing arguments", async () => {
  mockContext.userHasPermissionLegacy.mockReturnValue(true);
  mockShopById.mockReturnValue(fakeShop);

  await expect(enablePaymentMethodForShop(mockContext, {})).rejects.toThrowErrorMatchingSnapshot();
});

test("errors on invalid payment method", async () => {
  mockContext.userHasPermissionLegacy.mockReturnValue(true);
  mockShopById.mockReturnValue(fakeShop);

  await expect(enablePaymentMethodForShop(mockContext, {
    shopId: fakeShop._id,
    paymentMethodName: "does not exist",
    isEnabled: true
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("errors on invalid shop", async () => {
  mockContext.userHasPermissionLegacy.mockReturnValue(true);
  mockShopById.mockReturnValue();

  await expect(enablePaymentMethodForShop(mockContext, {
    shopId: "does not exist",
    paymentMethodName: "mockPaymentMethod",
    isEnabled: true
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("enables payment method for valid shop", async () => {
  fakeShop.availablePaymentMethods = ["mockPaymentMethod"];
  mockContext.userHasPermissionLegacy.mockReturnValue(true);
  mockShopById.mockReturnValue(fakeShop);
  mockPaymentMethods.mockReturnValue([{
    name: "mockPaymentMethod",
    displayName: "Mock!",
    pluginName: "mock-plugin",
    isEnabled: true
  }]);

  await expect(enablePaymentMethodForShop(mockContext, {
    shopId: fakeShop._id,
    paymentMethodName: "mockPaymentMethod",
    isEnabled: true
  })).resolves.toEqual([{
    name: "mockPaymentMethod",
    displayName: "Mock!",
    pluginName: "mock-plugin",
    isEnabled: true
  }]);

  expect(mockPaymentMethods).toHaveBeenCalledWith(mockContext, fakeShop._id);
});

test("disables payment method for valid shop", async () => {
  mockContext.checkPermissionsLegacy.mockReturnValueOnce(Promise.resolve(null));
  mockShopById.mockReturnValue(fakeShop);
  mockPaymentMethods.mockReturnValue([{
    name: "mockPaymentMethod",
    displayName: "Mock!",
    pluginName: "mock-plugin",
    isEnabled: false
  }]);

  await expect(enablePaymentMethodForShop(mockContext, {
    shopId: fakeShop._id,
    paymentMethodName: "mockPaymentMethod",
    isEnabled: false
  })).resolves.toEqual([{
    name: "mockPaymentMethod",
    displayName: "Mock!",
    pluginName: "mock-plugin",
    isEnabled: false
  }]);

  expect(mockPaymentMethods).toHaveBeenCalledWith(mockContext, fakeShop._id);
});
