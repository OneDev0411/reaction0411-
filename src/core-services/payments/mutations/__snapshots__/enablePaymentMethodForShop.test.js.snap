// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors on invalid payment method 1`] = `"checkPermissions is not a function"`;

exports[`errors on invalid shop 1`] = `"checkPermissions is not a function"`;

exports[`errors on missing arguments 1`] = `"Is enabled is required"`;

exports[`throws if userHasPermission returns false 1`] = `"checkPermissions is not a function"`;
